### PR TITLE
test: adds outbound eviction functional tests, updates comment in ConsiderEviction

### PR DESCRIPTION
**Pull request by sr-gi**
## Motivation

While checking the outbound eviction code I realized a case was not considered within the comments, which in turn made me realize we had no functional tests for the outbound eviction case (when I went to check/add the test case).

This PR updates the aforementioned comment and adds functional tests to cover the outbound eviction logic, in addition to the existing unit tests found at `src/test/denialofservice_tests.cpp`.

### CODE REVIEW

**Comment by sr-gi** [comment]
While the tests seem to run fine, I noticed a warning that appears every now and then:

```
TestFramework.p2p (WARNING): Connection lost to 0:0 due to [Errno 54] Connection reset by peer
```

I'm not sure if something unintended is happening under the hood. Disconnections are expected, but it's our node who is supposed to trigger them, so maybe I'm missing something.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -5098,9 +5098,10 @@ void PeerManagerImpl::ConsiderEviction(CNode& pto, Peer& peer, std::chrono::seco
                 state.m_chain_sync.m_sent_getheaders = false;
             }
         } else if (state.m_chain_sync.m_timeout == 0s || (state.m_chain_sync.m_work_header != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.m_chain_sync.m_work_header->nChainWork)) {
-            // Our best block known by this peer is behind our tip, and we're either noticing
-            // that for the first time, OR this peer was able to catch up to some earlier point
-            // where we checked against our tip.
+            // Either:
```
Comment (naumenkogs) [inline src/net_processing.cpp:5101]: I've spent a ridiculous amount of time understanding what's going on here, and I'm still far from being 100% sure. Perhaps if you touch this comment, you can expand it even further? E.g., connecting to each of the conditions in `else if`

Comment (sr-gi) [reply src/net_processing.cpp:5101]: Looks like I'm not the only one who took a minute to get this 😅 

I wouldn't mind expanding the comment if there is am overall agreement that it is incomplete

Comment (sr-gi) [reply src/net_processing.cpp]: Rebased and updated the comment. I also extended the test slightly to cover all possible cases

**Comment by sr-gi** [comment]
Rebased to fix CI

**Comment by sr-gi** [comment]
Rebased. This has failed CI once when waiting for the disconnection after:

```
cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
```
I've tried to reproduce it but I cannot in my local setup, the only thing that comes to mind is that mock time may be too close to the timeout (only one second ahead), so I gave it a delta of 10 seconds instead

File: test/functional/p2p_eviction.py [hunk 2]
```diff
@@ -122,6 +129,178 @@ def run_test(self):
...
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the peer does not get evicted")
+        test_node.sync_with_ping()
+
+        node.disconnect_p2ps()
+
+    def test_outbound_eviction_protected(self):
```
Comment (tdb3) [inline test/functional/p2p_eviction.py:222]: Added notes below to provide insight into what was reviewed:

In `test_outbound_eviction_protected()` if unexpected eviction occurs (i.e. the P2P connection is no longer present), then `sync_with_ping()` will raise (`IOError('Not connected')`, causing the test to fail.  Moved node.disconnect_p2ps() above the last test_node.sync_with_ping() call and, as expected, this occurred.

nit: The heading comment in the `test_outbound_eviction_protected()` omits that a protected peer is also non-block-only, but this is implicit in the call to add_outbound_p2p_connection() using outbound-full-relay as the connection_type.

net_processing.cpp:
```
A peer is marked as protected if all of these are true:
 - its connection type is IsBlockOnlyConn() == false
 - it gave us a valid connecting header
 - we haven't reached MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT yet
 - its chain tip has at least as much work as ours
```

File: test/functional/p2p_eviction.py [hunk 3]
```diff
@@ -122,6 +129,178 @@ def run_test(self):
...
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the node does not get evicted")
+        test_node.sync_with_ping()
+
+        node.disconnect_p2ps()
+
+    def test_outbound_eviction_protected_mixed(self):
```
Comment (tdb3) [inline test/functional/p2p_eviction.py:252]: nit:  I could be missing something, but a test case covering the limit of protected outbound peers (e.g. `MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT`) was not observed.  Do you think it's worth adding?  

Thinking out loud:
`MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT = 4`, so maybe adding another peer that would normally be considered protected (i.e. one that provided a block with the same amount of work as the tip) and seeing that it is not protected could test this?

**Review by tdb3 — COMMENTED** [review]
Excellent work finding a gap and filling it by increasing functional test robustness.
ACK for edc14bc9da332cae39f8803db559b532b3c74e16

Checked out the PR branch, built, and ran all functional tests (all passed).

The review was focused on verifying edge case coverage of the tests and that tests will accurately catch both pass and fail conditions.

Did not observe previously reported warning message:
```
TestFramework.p2p (WARNING): Connection lost to 0:0 due to [Errno 54] Connection reset by peer
```

File: src/net_processing.cpp [hunk 4]
```diff
@@ -5132,16 +5132,19 @@ void PeerManagerImpl::ConsiderEviction(CNode& pto, Peer& peer, std::chrono::seco
         // unless it's invalid, in which case we should find that out and
         // disconnect from them elsewhere).
         if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= m_chainman.ActiveChain().Tip()->nChainWork) {
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply src/net_processing.cpp:5134]: That's covered in the last part of `test_outbound_eviction_unprotected`

https://github.com/bitcoin/bitcoin/blob/edc14bc9da332cae39f8803db559b532b3c74e16/test/functional/p2p_eviction.py#L202-L218

File: test/functional/p2p_eviction.py (hunk 2 above)
Comment (sr-gi) [reply test/functional/p2p_eviction.py:222]: Protected peer actually implies `outbound-full-relay`. The comment you mentioned may be outdated. I may add an update to it to the PR.

https://github.com/bitcoin/bitcoin/blob/edc14bc9da332cae39f8803db559b532b3c74e16/src/net_processing.cpp#L2875-L2885

File: test/functional/p2p_eviction.py (hunk 3 above)
Comment (sr-gi) [reply test/functional/p2p_eviction.py:252]: The limit is tested in this test. We create 8 connections, from which only 4 are protected. If you mean that there are no unprotected peers that behave properly, that could be added. e.g.

4 protected, send initial headers, do nothing, don't get evicted
2 unprotected, keep in sync, don't get evicted
2 unprotected, do not keep up, get evicted

**Review by sr-gi — COMMENTED** [review]
Thanks for the review @tdb3

File: src/net_processing.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tdb3) [reply src/net_processing.cpp:5134]: Correct. >= rather than >.

**Comment by sr-gi** [comment]
I redefined how peers are split in `p2p_eviction:test_outbound_eviction_protected_mixed` to address https://github.com/bitcoin/bitcoin/pull/29122#discussion_r1485375517. Both commits can be squashed (leaving them as they are for now in case someone does not agree with the approach)

File: test/functional/p2p_eviction.py (hunk 3 above)
Comment (tdb3) [reply test/functional/p2p_eviction.py:252]: Thank you, yes, I should have been clearer.  The updated `test_outbound_eviction_protected_mixed` in 91005f7459993e1f7139e46b53eef1bb04860a7c looks great (includes 4 protected, 4 unprotected (2 honest, 2 misbehaving by sending no headers or old)).  ACK

**Comment by sr-gi** [comment]
> Rebased. This has failed CI once when waiting for the disconnection after:
> 
> ```
> cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
> ```
> 
> I've tried to reproduce it but I cannot in my local setup, the only thing that comes to mind is that mock time may be too close to the timeout (only one second ahead), so I gave it a delta of 10 seconds instead.
> 
> PS: For context, you can check the failure here: https://cirrus-ci.com/task/4624744360706048?logs=ci#L6520. See how the peer is given the final notice by sending a getheaders message, but the check is performed before the peer gets disconnected.

I've been looking more into this, given an error showed up again on CI. I think the issue arose from the initial mock time not being properly set. I've reset the increments to be just one second apart, plus set the initial mock time before the tests were run.

Also, I've moved the outbound eviction tests to their own file, given `p2p_eviction` was being described as testing inbound eviction, plus used some custom classes that were only relevant for inbounds.

**Comment by tdb3** [comment]
> > Rebased. This has failed CI once when waiting for the disconnection after:
> > ```
> > cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I've tried to reproduce it but I cannot in my local setup, the only thing that comes to mind is that mock time may be too close to the timeout (only one second ahead), so I gave it a delta of 10 seconds instead.
> > PS: For context, you can check the failure here: https://cirrus-ci.com/task/4624744360706048?logs=ci#L6520. See how the peer is given the final notice by sending a getheaders message, but the check is performed before the peer gets disconnected.
> 
> I've been looking more into this, given an error showed up again on CI. I think the issue arose from the initial mock time not being properly set. I've reset the increments to be just one second apart, plus set the initial mock time before the tests were run.
> 
> Also, I've moved the outbound eviction tests to their own file, given `p2p_eviction` was being described as testing inbound eviction, plus used some custom classes that were only relevant for inbounds.

(now non-stale) ACK for f06b50dfd45000f34bf258d48101718b4b5f9772.
Pulled.  Built.  Ran `p2p_eviction.py` and `p2p_outbound_eviction.py`.  Both passed.

**Review by cbergqvist — COMMENTED** [review]
Reviewed f06b50d. Appreciate the additional test coverage!

**Comment by cbergqvist** [comment]
Couldn't reproduce the CI timeout issue btw. But I have encountered the warning:
```
2024-02-22T22:01:47.354000Z TestFramework.p2p (WARNING): Connection lost to 0:0 due to [Errno 104] Connection reset by peer
```

File: src/net_processing.cpp [hunk 5]
```diff
@@ -5132,16 +5132,19 @@ void PeerManagerImpl::ConsiderEviction(CNode& pto, Peer& peer, std::chrono::seco
...
                 state.m_chain_sync.m_timeout = 0s;
                 state.m_chain_sync.m_work_header = nullptr;
                 state.m_chain_sync.m_sent_getheaders = false;
             }
-        } else if (state.m_chain_sync.m_timeout == 0s || (state.m_chain_sync.m_work_header != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.m_chain_sync.m_work_header->nChainWork)) {
-            // Our best block known by this peer is behind our tip, and we're either noticing
-            // that for the first time, OR this peer was able to catch up to some earlier point
-            // where we checked against our tip.
-            // Either way, set a new timeout based on current tip.
+        } else if (state.m_chain_sync.m_timeout == 0s || (state.m_chain_sync.m_work_header != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.m_chain_sync.m_work_header->nChainWork)) {\
```
Comment (marcofleon) [inline src/net_processing.cpp:5141]: nit: It looks like the backslash at the end of this line is a typo, yes? Could be removed for clarity.

**Review by marcofleon — COMMENTED** [review]
ACK. Well done on the new test. Seems like a good number of cases are included. Only thing I commented on was the backslash nit in `net_processing.cpp`. 

No issues or warnings when built and tested. Tests `p2p_eviction.py` and `p2p_outbound_eviction.py` ran successfully.

File: test/functional/p2p_outbound_eviction.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        # This should also apply for a node doing IBD that does not catch up in time
+        # Get our tip header and its parent
+        tip_header = from_hex(CBlockHeader(), node.getblockheader(node.getbestblockhash(), False))
+        prev_header = from_hex(CBlockHeader(), node.getblockheader(f"{tip_header.hashPrevBlock:064x}", False))
+
+        # Connect a peer and make it send us headers ending in our tip's parent
+        test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="outbound-full-relay")
+        headers_message = msg_headers()
+        headers_message.headers = [prev_header]
+        test_node.send_and_ping(headers_message)
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:67]: You're right, I didn't realize that the `msg_headers` constructor accepted an optional `headers` param

File: test/functional/p2p_outbound_eviction.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+        # Trigger the timeouts
+        cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:72]: From the `sync_with_ping` docs/comments:

```
# Sending two pings back-to-back, requires that the node calls
# `ProcessMessage` twice, and thus ensures `SendMessages` must have
# been called at least once
```

For the sake of this test, this means that we are triggering the outbound eviction checks given we mock the node time right before to be past the check timeout

File: test/functional/p2p_outbound_eviction.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:73]: That is already being tested implicitly by `sync_with_ping()` + wait + check for the disconnection.

By mocking a time passed `CHAIN_SYNC_TIMEOUT` and syncing (and not getting an error) we test that the connection has not been dropped, therefore `node` must have followed to send the get headers message.

I'll add it for clarity anyway

File: src/net_processing.cpp (hunk 5 above)
Comment (sr-gi) [reply src/net_processing.cpp:5141]: Nice catch

Comment (sr-gi) [reply src/net_processing.cpp:5141]: Nice catch!

**Comment by sr-gi** [comment]
Thanks for the reviews @cbergqvist and @marcofleon, I've addressed the comments of both.

I also squashed the two commits from @tdb3 given looks like no one ha opposed to those

File: src/net_processing.cpp (hunk 5 above)
Comment (cbergqvist) [reply src/net_processing.cpp]: I assumed it was some super-rare syntax for making tooling understand the comment below applied to the line before it. :D

File: test/functional/p2p_outbound_eviction.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cbergqvist) [reply test/functional/p2p_outbound_eviction.py:71]: Yeah, I saw that comment. Your explanation for the context of this test helps clarify it, thanks!

Calling "sync" (synchronize) on a bitcoin node and having it get the highest tip from it's peers would be less surprising. Maybe "update" or "process" would be better verbs but let's not rename it in this PR.

**Review by cbergqvist — APPROVED** [review]
ACK bffbb17!

(Diffed `test/functional/p2p_outbound_eviction.py` against previous implementation and ran it a couple of times).

**Comment by davidgumberg** [comment]
Nit: What do you think about testing that block-relay-only nodes get evicted and aren't eligible for protection?

```python
# Ensure that block-relay-only peers are not granted protection and ARE evicted
self.log.info("Create an outbound connection to a peer that shares our tip but is block-relay only, so it does not get protection")
test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=1, connection_type="block-relay-only")

test_node.send_and_ping(msg_headers([from_hex(CBlockHeader(), node.getblockheader(node.getbestblockhash(), False))]))

self.log.info("Mine a new block and sync with our peer")
self.generateblock(node, output="raw(42)", transactions=[])
test_node.sync_with_ping()

self.log.info("Let enough time pass for the timeouts to go off")
# Trigger the timeouts and check how we are still connected
cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
node.setmocktime(cur_mock_time)
test_node.sync_with_ping()
test_node.wait_for_getheaders()
cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
node.setmocktime(cur_mock_time)
self.log.info("Test that the block-relay-only peer gets evicted")
test_node.wait_for_disconnect()
```

Otherwise, reviewed `ConsiderEviction` and the tests added here provide good coverage. I ran the `p2p_outbound_eviction.py` tests and they passed, I was unable to reproduce the connection reset mentioned above.

ACK https://github.com/bitcoin/bitcoin/pull/29122/commits/bffbb176d8a4f545c26ff7e437ab33cb004852bb

**Comment by tdb3** [comment]
re ACK bffbb176d8a4f545c26ff7e437ab33cb004852bb

**Comment by sr-gi** [comment]
I've added a test to cover for @tdb3 and @davidgumberg comments on non-outbound-full-relay peers eviction, plus amended the original commit to rename `test_outbound_eviction_mixed` and added a missing comment to the method.

File: test/functional/p2p_outbound_eviction.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
+        test_node.wait_for_getheaders()
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the peer gets evicted")
+        test_node.wait_for_disconnect()
+
+
+    def run_test(self):
```
Comment (marcofleon) [inline test/functional/p2p_outbound_eviction.py:238]: I could be missing something here but the new test `test_outbound_eviction_blocks_relay_only` isn't included in `run_test`. It does look good from just reading it though :)

Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:238]: You're right, this may have been removed while rebasing. Nice catch!

This should be fixed now

**Review by marcofleon — APPROVED** [review]
re-ACK 33353b7723d0704d64c0dfc1b0f5d072692e68b9. I've reviewed the new changes, focusing on the added negative test. I think it's a valuable addition, expanding the test's coverage of different scenarios in the network. I built and ran `p2p_outbound_eviction` multiple times and observed no issues or warnings.

Well done on addressing the feedback and enhancing the test suite.

File: test/functional/p2p_outbound_eviction.py [hunk 10]
```diff
@@ -207,12 +207,40 @@ def test_outbound_eviction_mixed(self):
         for peer in misbehaving_unprotected_peers:
             peer.wait_for_disconnect()
 
+        node.disconnect_p2ps()
```
Comment (davidgumberg) [inline test/functional/p2p_outbound_eviction.py:210]: Annoying nit: This should be in the previous commit (feel free to disregard)

**Review by cbergqvist — APPROVED** [review]
ACK 33353b7.

Diffed `p2p_outbound_eviction.py` against previously acked  version.

Ran the new version a few times, switched the node type from `block-relay-only` to `outbound-full-relay` and confirmed that `wait_for_disconnect()` times out  (decreased timeout to 10s locally, but should still be more than enough).

**Comment by davidgumberg** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/29122/commits/33353b7723d0704d64c0dfc1b0f5d072692e68b9

Tested by modifying `bool IsOutboundOrBlockRelayConn()`, and `CHAIN_SYNC_TIMEOUT` and this test catches. It would be nice if we also had a test that made sure we don't punish peers and set the timeout to `HEADERS_RESPONSE_TIME` *before* we've given them `CHAIN_SYNC_TIMEOUT` to catch up, but I think this PR looks great as it stands!

**Comment by sr-gi** [comment]
Rebased to deal with merge conflicts

**Comment by davidgumberg** [comment]
reACK https://github.com/bitcoin/bitcoin/commit/59affd0bf9042d746a6b32739edb0cd45d28d569

**Comment by tdb3** [comment]
re ACK for 59affd0bf9042d746a6b32739edb0cd45d28d569.
Built and re-ran functional tests locally.  All passed.

**Review by cbergqvist — APPROVED** [review]
re ACK 59affd0bf9042d746a6b32739edb0cd45d28d569

Built & ran `test/functional/test_runner.py --cachedir=/mnt/tmp/cache --tmpdir=/mnt/tmp` successfully.

**Review by marcofleon — APPROVED** [review]
re-ACK 59affd0bf9042d746a6b32739edb0cd45d28d569. Ran all functional tests, no issues.

**Comment by sr-gi** [comment]
Rebased to fix CI due to https://github.com/bitcoin/bitcoin/pull/29610

**Comment by davidgumberg** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/29122/commits/cd5b7a11ee55b1f86bb7662c4fc83616ffa735b1

```console
$ git range-diff a945f09...59affd0 015ac13...cd5b7a1                                              
1:  a12a4dfa90 = 1:  8fa272dccc test: adds outbound eviction functional tests, updates comment in ConsiderEviction     
2:  59affd0bf9 = 2:  cd5b7a11ee test: adds outbound eviction tests for non outbound-full-relay peers 

$ python ./test/functional/test_runner.py
ALL                                                    | ✓ Passed  | 1350 s (accumulated) 
```

**Review by cbergqvist — APPROVED** [review]
reACK cd5b7a11ee55b1f86bb7662c4fc83616ffa735b1

(Diff didn't change). Built & ran functional tests.

**Review by tdb3 — COMMENTED** [review]
re ACK for cd5b7a11ee55b1f86bb7662c4fc83616ffa735b1.
Built and re-ran functional tests locally. All passed.

**Review by marcofleon — COMMENTED** [review]
re-ACK cd5b7a11ee55b1f86bb7662c4fc83616ffa735b1. All functional tests passed.

File: test/functional/p2p_outbound_eviction.py (hunk 10 above)
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:210]: I don't think that's the case. In the previous commit this was the last test so there was no need to manually disconnect the nodes. This is only needed after the last commit to start fresh

**Review by stratospher — COMMENTED** [review]
cool test! Concept ACK. left some suggestions, nits and 1 problem.

something I found interesting was when `generateblock`  is called - `inv`, `getdata` and `block` messages are exchanged between node and peer. so in the tests (ex: `test_outbound_eviction_unprotected`), the peer is sending the header of an older block it already knows about! I don't think it affects the behaviour in the tests. but would like to make sure that this is expected.

File: test/functional/p2p_outbound_eviction.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    from_hex,
+    msg_headers,
+    CBlockHeader,
+)
+from test_framework.p2p import (
+    P2PInterface,
+)
+from test_framework.test_framework import BitcoinTestFramework
+
+# Timeouts (in minutes)
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:28]: I'll change it for seconds 🤦

File: test/functional/p2p_outbound_eviction.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        self.num_nodes = 1
+
+    def test_outbound_eviction_unprotected(self):
+        # This tests the eviction logic for **unprotected** outbound peers (that is, PeerManagerImpl::ConsiderEviction)
+        node = self.nodes[0]
+        cur_mock_time = node.mocktime
+
+        self.log.info("Create an outbound connection and don't send any headers")
+        # Test disconnect due to no block being announced in 22+ minutes (headers are not even exchanged)
+        test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="outbound-full-relay")
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:44]: Yeah, I see your point. The rationale was that it was the node being tested, but that's indeed confusing given `TestNode`, I'll rename it to peer.

File: test/functional/p2p_outbound_eviction.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+        # Wait for over 20 min to trigger the first eviction timeout. This sets the last call past 2 min in the future.
+        test_node.sync_with_ping()
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:46]: I don't think we do

File: test/functional/p2p_outbound_eviction.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+            best_block_hash = self.generateblock(node, output="raw(42)", transactions=[])["hash"]
+            cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
+            node.setmocktime(cur_mock_time)
+
+            # Send a header with the old tip
+            test_node.send_and_ping(msg_headers([tip_header]))
+            tip_header = from_hex(CBlockHeader(), node.getblockheader(best_block_hash, False))
+
+            # Check that we are not evicted
+            test_node.wait_for_getheaders()
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:99]: Ohh, TIL, I'm guessing this is because the collection is already populated with the `getheaders` from the interaction `i-1` and we find it all the time

File: test/functional/p2p_outbound_eviction.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_outbound_eviction.py:46]: [this line](https://github.com/bitcoin/bitcoin/blob/c1223188e0a5fb11c3a1b9224511a49dc2f848ed/test/functional/test_framework/test_node.py#L787) - but not a blocker, maybe the test is more consistent to read this way.

File: test/functional/p2p_outbound_eviction.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_outbound_eviction.py:99]: yeah! follow up idea, out of scope for this PR - it'd be awesome to modify `wait_for_getheaders()` to make sure that it's the same getheaders (which was sent from TestNode's `MaybeSendGetHeaders`) we're expecting that is received!

[the remark this answers is not in the document]
Comment (stratospher) [reply test/functional/p2p_outbound_eviction.py:99]: oh see #29318.

[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py]: Fixed it by popping from `last_message`, may open a followup simplifying it after #https://github.com/bitcoin/bitcoin/issues/18614 gets fixed

**Comment by sr-gi** [comment]
Rebased + addressed comments from @stratospher. I had to slightly reshape `test_outbound_eviction_unprotected` to fit https://github.com/bitcoin/bitcoin/pull/29122#discussion_r1535542317 given some time mocks + waits were unnecessary

File: test/functional/p2p_outbound_eviction.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+        self.num_nodes = 1
+
+    def test_outbound_eviction_unprotected(self):
+        # This tests the eviction logic for **unprotected** outbound peers (that is, PeerManagerImpl::ConsiderEviction)
+        node = self.nodes[0]
+        cur_mock_time = node.mocktime
+
+        self.log.info("Create an outbound connection and don't send any headers")
+        # Test disconnect due to no block being announced in 22+ minutes (headers are not even exchanged)
+        peer = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="outbound-full-relay")
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:44]: Yeah, I noticed that while patching your last comments for `test_outbound_eviction_unprotected` (which already has that change applied). I'm assessing how easy would be to supersede #29318 so this PR does not become super verbose and hacky

[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:44]: #29736

Will rebase on top of that one once it gets merged

[the remark this answers is not in the document]
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:44]: I patched this locally on top of the aforementioned PR and looks like all issues are gone :)

**Comment by tdb3** [comment]
reACK for 368389d973a85d21aa9ebf030ef5e7f7ac5343de.
Pulled, built, ran all unit/functional tests (all passed).

File: test/functional/p2p_outbound_eviction.py [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+        node.disconnect_p2ps()
+
+    def test_outbound_eviction_blocks_relay_only(self):
+        # The logic for outbound eviction protection only applies to outbound-full-relay peers
+        # This tests that other types of peers (blocks-relay-only for instance) are not granted protection
+        node = self.nodes[0]
+        cur_mock_time = node.mocktime
+
+        self.log.info("Create an blocks-only outbound connection to a peer that shares our tip. This would usually grant protection")
+        test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="block-relay-only")
```
Comment (cbergqvist) [inline test/functional/p2p_outbound_eviction.py:222]: nit: Noticed you renamed `test_node` -> `peer` above (based off https://github.com/bitcoin/bitcoin/pull/29122#discussion_r1531465214) but not inside `test_outbound_eviction_blocks_relay_only()`.

(Was diffing against older head (59affd0bf9042d746a6b32739edb0cd45d28d569)).

Might be worth changing once you rebase on top of #29736.

Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:223]: Nice catch!

**Comment by sr-gi** [comment]
Rebased on top of https://github.com/bitcoin/bitcoin/pull/29736. Now this is using `wait_for_getheaders` to reduce the boilerplate of having to manually pop the results from `last_message`.

cc/ @stratospher

File: test/functional/p2p_outbound_eviction.py (hunk 10 above)
Comment (davidgumberg) [reply test/functional/p2p_outbound_eviction.py:213]: On second thought it seems a bad idea to depend on the order in which tests are called for deciding whether or not to perform cleanup. I think that might be kind of a footgun for anyone modifying the test in the future

**Comment by davidgumberg** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/29122/commits/be2c5510521081119f62ef3b29f76bb9097d0f34
Re-reviewed code, and attempted to break tests by making `ConsiderEviction` misbehave.

It is worth noting that these tests will not catch if outbound eviction logic gets triggered too early. e.g.:

```cpp
/** How long to wait for a peer to respond to a getheaders request */
static constexpr auto HEADERS_RESPONSE_TIME{0s};
// [...]
/** Timeout for (unprotected) outbound peers to sync to our chainwork */
static constexpr auto CHAIN_SYNC_TIMEOUT{0s};
```
 will pass. That seems fine to me, maybe something for a follow up.

File: test/functional/p2p_outbound_eviction.py (hunk 10 above)
Comment (sr-gi) [reply test/functional/p2p_outbound_eviction.py:213]: Ok, agreed

**Comment by sr-gi** [comment]
> ACK [be2c551](https://github.com/bitcoin/bitcoin/commit/be2c5510521081119f62ef3b29f76bb9097d0f34) Re-reviewed code, and attempted to break tests by making `ConsiderEviction` misbehave.
> 
> It is worth noting that these tests will not catch if outbound eviction logic gets triggered too early. e.g.:
> 
> ```c++
> /** How long to wait for a peer to respond to a getheaders request */
> static constexpr auto HEADERS_RESPONSE_TIME{0s};
> // [...]
> /** Timeout for (unprotected) outbound peers to sync to our chainwork */
> static constexpr auto CHAIN_SYNC_TIMEOUT{0s};
> ```
> 
> will pass. That seems fine to me, maybe something for a follow-up.

I don't think I follow you here. `HEADERS_RESPONSE_TIME` and `CHAIN_SYNC_TIMEOUT` are constants hardcoded in `net_processing.cpp` (and mapped here). I would expect unit test to fail if those are updated.

**Comment by sr-gi** [comment]
Updated to address https://github.com/bitcoin/bitcoin/pull/29122#discussion_r1580221364

**Comment by davidgumberg** [comment]
reACK https://github.com/bitcoin/bitcoin/commit/d53d84834747c37f4060a9ef379e0a6b50155246

> I don't think I follow you here. `HEADERS_RESPONSE_TIME` and `CHAIN_SYNC_TIMEOUT` are constants hardcoded in `net_processing.cpp` (and mapped here). I would expect unit test to fail if those are updated.

I didn't mean to say that these tests should be responsible for enforcing the particular values of `HEADERS_RESPONSE_TIME` and `CHAIN_SYNC_TIMEOUT`, just that they would not catch a regression where outbound eviction logic gets triggered earlier than expected on peers, which can be demonstrated by setting both values to zero.

Anyways, I think if what I'm saying isn't confused in some way, it is more pedantic than useful.

**Review by cbergqvist — APPROVED** [review]
ACK d53d84834747c37f4060a9ef379e0a6b50155246

Functional including `--extended` tests passed (skipped unrelated `feature_dbcrash`).

**Comment by tdb3** [comment]
Re ACK for d53d84834747c37f4060a9ef379e0a6b50155246
Pulled, built, ran all unit/functional tests (all passed).

**Comment by Ferrydh** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Review by marcofleon — COMMENTED** [review]
Re ACK. Reviewed the code and the tests look good to me. I ran `p2p_outbound_eviction.py` individually and along with all the other functional tests and everything looks good.

**Comment by achow101** [comment]
ACK d53d84834747c37f4060a9ef379e0a6b50155246
